### PR TITLE
[staging] Multilanguage: Display an error when url language code is saved as empty.

### PR DESCRIPTION
**Pull request by infograf768**
### Summary of Changes

This PR prevents  saving a Content Language when its `Url Language Code` is empty.

### Testing Instructions
2 cases when editing or creating a Content Language

1. If only symbols like `%` or `&` are entered in the `URL Language Code` field

2. If multibytes characters are entered when `Unicode Alias` is not set to `Yes` in Global Configuration. 
Example while admin language is en-GB, the Greek characters `αω`

### Actual result BEFORE applying this Pull Request
The Content Language is saved with an empty Url Language Code

Screen Shot 2020-08-27 at 14 54 51

### Expected result AFTER applying this Pull Request
We now get an error message.

Screen Shot 2020-08-27 at 14 57 51
`

Screen Shot 2020-08-27 at 14 58 25

### CODE REVIEW

File: administrator/language/en-GB/en-GB.com_languages.ini [hunk 1]
```diff
@@ -7,9 +7,10 @@ COM_LANGUAGES="Languages"
 COM_LANGUAGES_CONFIGURATION="Languages: Options"
 COM_LANGUAGES_ERR_DELETE="Select a language to delete"
 COM_LANGUAGES_ERR_NO_LANGUAGE_SELECTED="No language selected."
+COM_LANGUAGES_ERR_PUBLISH="Select a language to publish."
 COM_LANGUAGES_ERROR_LANG_TAG="<br />The Language Tag should have 2 or 3 lowercase letters corresponding to the ISO language, a dash and 2 uppercase letters corresponding to the ISO country code. <br />This should be the exact prefix used for the language installed or to be installed. Example: en-GB, srp-ME."
 COM_LANGUAGES_ERROR_LANGUAGE_METAFILE_MISSING="Could not load %s language meta XML file from %s."
-COM_LANGUAGES_ERR_PUBLISH="Select a language to publish."
+COM_LANGUAGES_ERROR_SEF="The \"Url Language Code\" can't be saved. Please use only alphanumeric characters."
```
Comment (SharkyKZ) [inline administrator/language/en-GB/en-GB.com_languages.ini]:
```suggestion
COM_LANGUAGES_ERROR_SEF="The \"URL Language Code\" can't be saved. Please use only alphanumeric characters."
```

Comment (infograf768) [reply administrator/language/en-GB/en-GB.com_languages.ini]: Done. Thanks.

**Comment by sandewt** [comment]
I have tested this item :white_check_mark: successfully on 0f53eca651bc6956bc2a6312e05afbac0f5f2a60<br /><br />Joomla! 3.9.22-dev Development [ Amani ] 26-August-2020 08:12 GMT

For the sake of completeness:
A horizontal dash (-) is accepted too. For example ```en-gb```. The - is not an alphanumeric character.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30496">issues.joomla.org/tracker/joomla-cms/30496</a>.</sub>

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on 0f53eca651bc6956bc2a6312e05afbac0f5f2a60<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30496">issues.joomla.org/tracker/joomla-cms/30496</a>.</sub>

**Comment by brianteeman** [comment]
Shouldn't this be done as a validation rule in the xml ?

**Comment by brianteeman** [comment]
@sandewt that means that at least part of this pr is incorrect

**Comment by infograf768** [comment]
> A horizontal dash (-) is accepted too. For example en-gb. The - is not an alphanumeric character.

Yep. The hyphen is a accepted, but only if used between alphanumeric characters. Any idea for the value of the string to reflect this?

**Comment by SharkyKZ** [comment]
Maybe use the same text as for aliases?

**Comment by infograf768** [comment]
As for https://github.com/joomla/joomla-cms/pull/30497, I explained there that we can use multibytes characters when `Unicode Alias` is set to YES in Global Configuration. It does not break anything, this is why this is specifid in the test instructions.

@SharkyKZ Looking now what we have for alias.

**Comment by infograf768** [comment]
Is'nt the alias hint a bit too long and overly complex here?

`JFIELD_ALIAS_DESC="The Alias will be used in the SEF URL. Leave this blank and Joomla will fill in a default value from the title. This value will depend on the SEO settings (Global Configuration->Site). <br />Using Unicode will produce UTF-8 aliases. You may also enter manually any UTF-8 character. Spaces and some forbidden characters will be changed to hyphens.<br />When using default transliteration it will produce an alias in lower case and with dashes instead of spaces. You may enter the Alias manually. Use lowercase letters and hyphens (-). No spaces or underscores are allowed. Default value will be a date and time if the title is typed in non-latin letters."`

In our case, as we throw the error only when the field remains empty when saved — and `Unicode Alias` is set to NO —, maybe we could simply use

`COM_LANGUAGES_ERROR_SEF="The \"Url Language Code\" can't be saved. Please use only alphanumeric characters and hyphens (-)."`

Shall we add that entering an hyphen alone will still trigger the error?

**Comment by SharkyKZ** [comment]
I think the part about unicode aliases should be added.

**Comment by infograf768** [comment]
Something like ?
`COM_LANGUAGES_ERROR_SEF="The \"Url Language Code\" can't be saved.<br>Please use only alphanumeric characters and hyphens (-). Or, if you want to use UTF8 characters, make sure that \"Unicode Alias\" is set to \"Yes\" in Global Configuration."`

Screen Shot 2020-08-28 at 07 51 01

**Comment by brianteeman** [comment]
`COM_LANGUAGES_ERROR_SEF="The \"Url Language Code\" can't be saved.<br>Please use only alphanumeric characters and hyphens (-).  \"Unicode Alias\" should be set to \"Yes\" in Global Configuration to use UTF8 characters"
`

**Comment by sandewt** [comment]
Quite difficult to understand all of this. I found the following explanation:
[https://docs.joomla.org/International_Enhancements_for_Version_1.6#Unicode_URLs](https://docs.joomla.org/International_Enhancements_for_Version_1.6#Unicode_URLs)

**Comment by infograf768** [comment]
@brianteeman 
Good proposal. Modifying now.

File: administrator/language/en-GB/en-GB.com_languages.ini [hunk 2]
```diff
@@ -10,7 +10,7 @@ COM_LANGUAGES_ERR_NO_LANGUAGE_SELECTED="No language selected."
 COM_LANGUAGES_ERR_PUBLISH="Select a language to publish."
 COM_LANGUAGES_ERROR_LANG_TAG="<br />The Language Tag should have 2 or 3 lowercase letters corresponding to the ISO language, a dash and 2 uppercase letters corresponding to the ISO country code. <br />This should be the exact prefix used for the language installed or to be installed. Example: en-GB, srp-ME."
 COM_LANGUAGES_ERROR_LANGUAGE_METAFILE_MISSING="Could not load %s language meta XML file from %s."
-COM_LANGUAGES_ERROR_SEF="The \"URL Language Code\" can't be saved. Please use only alphanumeric characters."
+COM_LANGUAGES_ERROR_SEF="The \"Url Language Code\" can't be saved.<br>Please use only alphanumeric characters and hyphens (-). \"Unicode Alias\" should be set to \"Yes\" in Global Configuration to use UTF8 characters."
```
Comment (SharkyKZ) [inline administrator/language/en-GB/en-GB.com_languages.ini]:
```suggestion
COM_LANGUAGES_ERROR_SEF="The \"URL Language Code\" can't be saved.<br>Please use only alphanumeric characters and hyphens (-). \"Unicode Alias\" should be set to \"Yes\" in Global Configuration to use UTF-8 characters."
```

Comment (infograf768) [reply administrator/language/en-GB/en-GB.com_languages.ini]: Right. One should never copy/paste without being sure. ;)

Comment (infograf768) [reply administrator/language/en-GB/en-GB.com_languages.ini]: Done

**Comment by infograf768** [comment]
> `<br>` should be `<br />`

Not necessary and these are also globally modified in J4 strings.

**Comment by sandewt** [comment]
>Not necessary and these are also globally modified in J4 strings.

The text in the manual should then be adjusted, see:

[https://github.com/joomla/coding-standards/blob/master/manual/html.md#adding-line-breaks](https://github.com/joomla/coding-standards/blob/master/manual/html.md#adding-line-breaks)

**Comment by infograf768** [comment]
It is confusing indeed.
https://www.tutorialspoint.com/What-is-the-correct-way-of-using-br-br-or-br-in-HTML

In J4 we have 
9 occurences of `<br />`
216 occurences of `<br>`

**Comment by sandewt** [comment]
>It is confusing indeed.

Who is authorized to change the manual for J4?

**Comment by infograf768** [comment]
I posted an issue and informed maintainers.

As the string value has changed, this needs new tests.

**Comment by brianteeman** [comment]
In Joomla 3 it should be <br />
In Joomla 4 is should be <br>

This is because joomla 4 is set to be html5 and joomla 3 is not - its an option

File: administrator/language/en-GB/en-GB.com_languages.ini [hunk 3]
```diff
@@ -7,9 +7,10 @@ COM_LANGUAGES="Languages"
 COM_LANGUAGES_CONFIGURATION="Languages: Options"
 COM_LANGUAGES_ERR_DELETE="Select a language to delete"
 COM_LANGUAGES_ERR_NO_LANGUAGE_SELECTED="No language selected."
+COM_LANGUAGES_ERR_PUBLISH="Select a language to publish."
 COM_LANGUAGES_ERROR_LANG_TAG="<br />The Language Tag should have 2 or 3 lowercase letters corresponding to the ISO language, a dash and 2 uppercase letters corresponding to the ISO country code. <br />This should be the exact prefix used for the language installed or to be installed. Example: en-GB, srp-ME."
 COM_LANGUAGES_ERROR_LANGUAGE_METAFILE_MISSING="Could not load %s language meta XML file from %s."
-COM_LANGUAGES_ERR_PUBLISH="Select a language to publish."
+COM_LANGUAGES_ERROR_SEF="The \"URL Language Code\" can't be saved.<br>Please use only alphanumeric characters and hyphens (-). \"Unicode Alias\" should be set to \"Yes\" in Global Configuration to use UTF-8 characters."
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.com_languages.ini]: Remove first sentence as it is redundant to `Save failed with the following error:`.
Maybe make it similar to line 11 and be consistent with the use of either `dash` or `hyphen`.

**Comment by infograf768** [comment]
Changed again string value (last time).
It is now
`COM_LANGUAGES_ERROR_SEF="<br />The \"URL Language Code\" should have only alphanumeric characters and dashes (-).<br />To use UTF-8 characters, \"Unicode Alias\" should be set to \"Yes\" in Global Configuration."`

Which gives

Screen Shot 2020-08-28 at 21 21 26

**Comment by Quy** [comment]
Simplify further?
> To use UTF-8 characters, set \"Unicode Alias\" to \"Yes\" in Global Configuration.

**Comment by jiweigert** [comment]
Just to be devil's advokate here:
Based on the hint, "de-" / "-de"  is a valid language code here?

I know it's not, so there should be in the hint included at least one example of the correct use of the dash like "de-DE".
Just for the "least unexperienced user / editor"...<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30496">issues.joomla.org/tracker/joomla-cms/30496</a>.</sub>

**Comment by infograf768** [comment]
Folks, I really start to be fed up with this PR...

**Comment by jiweigert** [comment]
Sorry that you start to fed up.

In my case, I'm simply not able to access the online help at all on my hosters installation.
That may a configuration error or the hoster simply removed the online help...
Anyway.. having at least one (or two) working examples of what should be filled in there,
is just good behaviour in case of an error... but thats just my opinion.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30496">issues.joomla.org/tracker/joomla-cms/30496</a>.</sub>

**Comment by infograf768** [comment]
@jiweigert
if you enter `-de` or `de-` it will automatically be saved as `de` and will not trigger an error.

**Comment by infograf768** [comment]
RTC as we can debate many months about a slightly different value.

Thanks for testing<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30496">issues.joomla.org/tracker/joomla-cms/30496</a>.</sub>

**Comment by HLeithner** [comment]
I'm happy with it, thanks. But another PR is of course welcome.

**Pull request merged**